### PR TITLE
Basic comments handling

### PR DESCRIPTION
**Pull request by ludwigjossieaux**
Hello,

Related to issue #271, please find a basic handling of the comments at step level. 

Based on the location (line, column) provided by the parser, the comments above the steps are considered commenting the step below. There is some edge cases not yet taken into account (comments below the last step, for instance); if you agree with the basic principle, I can make sure these are also taken into account. 

There are also few tests checking the feature is working for the output formats.

### CODE REVIEW

**Comment by dirkrombauts** [comment]
Cool, I like it. I had a quick review of the code and I think the overall direction is good. So please go ahead and add handling for the edge cases you mentioned.

One concern - it's a bit of a mystery to me: the continuous build successfully built your pull request, but when I checked it out locally and ran the tests, two of them were failing. Both of them in the FormattingAFeature file. The reason is that the expected JSON did not include arrays for the comments. Could you check that in your version please!

And again, thanks for this big contribution. It would be cool if you could finalize it this week so it can be released in time for the CukeUP conference next week :-)

**Comment by ludwigjossieaux** [comment]
Thanks for the feedback ! it's doable this week, I'll add the missing commits.

Concerning the two tests, you are right it is missing the comment arrays on the step object, it will be added with the next commits. However these tests were already failing with the fork I did (still the head revision on develop). It looks like the "Summary" object is missing between "Features" and "Configuration". Is it correct ?

**Comment by dirkrombauts** [comment]
I fixed the failing tests (by removing the "configuration" line which was an overspecification of the test). It's a bit of a mystery why the test sometimes passes and sometimes fails. I have my suspicions but no proof :-/

Anyway, you can now pull the latest commit from picklesdoc/develop with the fixed test.

**Comment by ludwigjossieaux** [comment]
Hi,
I've pushed the commits for the comments located after the last step of a scenario. I'm still not entirely sure how handle more freely placed comments (outside of steps). I've also added some more tests and fixed those not passing.
Let me know if would think of something else I could add

File: src/Pickles/Pickles.ObjectModel/ObjectModel/Comment.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    public class Comment
+    {
+        public Comment()
+        {
+            // Set default
+            this.Type = CommentType.Normal;
+        }
+
```
Comment (dirkrombauts) [inline src/Pickles/Pickles.ObjectModel/ObjectModel/Comment.cs]: This line contains whitespace, please remove it

File: src/Pickles/Pickles.ObjectModel/ObjectModel/Comment.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//  </copyright>
+//  --------------------------------------------------------------------------------------------------------------------
+
+using System.Collections.Generic;
```
Comment (dirkrombauts) [inline src/Pickles/Pickles.ObjectModel/ObjectModel/Comment.cs]: please remove the unused using

File: src/Pickles/Pickles.ObjectModel/ObjectModel/Location.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//  </copyright>
+//  --------------------------------------------------------------------------------------------------------------------
+
+using System.Collections.Generic;
```
Comment (dirkrombauts) [inline src/Pickles/Pickles.ObjectModel/ObjectModel/Location.cs]: Please remove the unused using.

File: src/Pickles/Pickles.ObjectModel/ObjectModel/Scenario.cs [hunk 4]
```diff
@@ -44,6 +44,8 @@ public Scenario()
 
         public Feature Feature { get; set; }
 
+        public Location Location { get; set; }
+
         #endregion
```
Comment (dirkrombauts) [inline src/Pickles/Pickles.ObjectModel/ObjectModel/Scenario.cs]: This is not strictly your code, but do me a favour and remove the `#region` from the class. Thanks!

File: src/Pickles/Pickles.ObjectModel/ObjectModel/Step.cs [hunk 5]
```diff
@@ -31,5 +38,9 @@ public class Step
         public Table TableArgument { get; set; }
 
         public string DocStringArgument { get; set; }
+
+        public Location Location { get; set; }
+
+        public List<Comment> Comments { get; set; }
```
Comment (dirkrombauts) [inline src/Pickles/Pickles.ObjectModel/ObjectModel/Step.cs]: Please remove the extra whitespace at the end of the `Comments { get; set; }` line

File: src/Pickles/Pickles.Test/DocumentationBuilders/Excel/WhenAddingAStepToAWorksheet.cs [hunk 6]
```diff
@@ -47,5 +49,107 @@ public void ThenStepAddedSuccessfully()
                 Check.That(worksheet.Cell("D5").Value).IsEqualTo(step.Name);
             }
         }
+
+        [Test]
+        public void ThenStepCommentsAreAddedSuccesfully()
```
Comment (dirkrombauts) [inline src/Pickles/Pickles.Test/DocumentationBuilders/Excel/WhenAddingAStepToAWorksheet.cs]: Successfully is spelled with three s in total :-)

File: src/Pickles/Pickles.Test/DocumentationBuilders/Excel/WhenAddingAStepToAWorksheet.cs [hunk 7]
```diff
@@ -47,5 +49,107 @@ public void ThenStepAddedSuccessfully()
...
+                excelStepFormatter.Format(worksheet, step, ref row);
+
+                Check.That(worksheet.Cell("C5").Value).IsEqualTo(step.Comments.First().Text);
+                Check.That(worksheet.Cell("C6").Value).IsEqualTo(step.NativeKeyword);
+                Check.That(worksheet.Cell("D6").Value).IsEqualTo(step.Name);
+            }
+        }
+
+        [Test]
+        public void ThenMultilineStepCommentsAreAddedSuccesfully()
```
Comment (dirkrombauts) [inline src/Pickles/Pickles.Test/DocumentationBuilders/Excel/WhenAddingAStepToAWorksheet.cs]: Successfully is spelled with three s in total :-)

File: src/Pickles/Pickles.Test/DocumentationBuilders/Excel/WhenAddingAStepToAWorksheet.cs [hunk 8]
```diff
@@ -47,5 +49,107 @@ public void ThenStepAddedSuccessfully()
...
+
+                Check.That(worksheet.Cell("C5").Value).IsEqualTo(step.Comments[0].Text);
+                Check.That(worksheet.Cell("C6").Value).IsEqualTo(step.Comments[1].Text);
+                Check.That(worksheet.Cell("C7").Value).IsEqualTo(step.NativeKeyword);
+                Check.That(worksheet.Cell("D7").Value).IsEqualTo(step.Name);
+            }
+        }
+
+        [Test]
+        public void ThenCommentsAfterTheLastStepAreAddedSuccesfully()
```
Comment (dirkrombauts) [inline src/Pickles/Pickles.Test/DocumentationBuilders/Excel/WhenAddingAStepToAWorksheet.cs]: Successfully is spelled with three s in total :-)

File: src/Pickles/Pickles.Test/DocumentationBuilders/HTML/WhenFormattingStep.cs [hunk 9]
```diff
@@ -181,5 +181,123 @@ public void Tables_are_formatted_as_list_items_with_tables_internal()
...
+        public void Comments_are_displayed_above_their_related_step()
+        {
+            var step = new Step
+            {
+                Keyword = Keyword.Given,
+                NativeKeyword = "Given ",
+                Name = "a simple step",
+                TableArgument = null,
+                DocStringArgument = null,
+                Comments = new List<Comment>()
```
Comment (dirkrombauts) [inline src/Pickles/Pickles.Test/DocumentationBuilders/HTML/WhenFormattingStep.cs]: the `()` is unnecessary.

File: src/Pickles/Pickles.Test/DocumentationBuilders/HTML/WhenFormattingStep.cs [hunk 10]
```diff
@@ -181,5 +181,123 @@ public void Tables_are_formatted_as_list_items_with_tables_internal()
...
+                {
+                    new Comment()
```
Comment (dirkrombauts) [inline src/Pickles/Pickles.Test/DocumentationBuilders/HTML/WhenFormattingStep.cs]: the `()` is unnecessary.

File: src/Pickles/Pickles.Test/DocumentationBuilders/HTML/WhenFormattingStep.cs [hunk 11]
```diff
@@ -181,5 +181,123 @@ public void Tables_are_formatted_as_list_items_with_tables_internal()
...
+                        Text = "    # A simple comment",
+                        Type = CommentType.StepComment
+                    }
+                }
+            };
+
+            var formatter = Container.Resolve<HtmlStepFormatter>();
+            XElement actual = formatter.Format(step);
+
+            XNamespace xmlns = XNamespace.Get("http://www.w3.org/1999/xhtml");
```
Comment (dirkrombauts) [inline src/Pickles/Pickles.Test/DocumentationBuilders/HTML/WhenFormattingStep.cs]: Do me a favor and introduce a private field with the namespace and use it in all tests in this class. Thanks!

File: src/Pickles/Pickles.Test/DocumentationBuilders/HTML/WhenFormattingStep.cs [hunk 12]
```diff
@@ -181,5 +181,123 @@ public void Tables_are_formatted_as_list_items_with_tables_internal()
...
+        public void Comments_after_the_last_step_are_displayed()
+        {
+            var step = new Step
+            {
+                Keyword = Keyword.Given,
+                NativeKeyword = "Given ",
+                Name = "a simple step",
+                TableArgument = null,
+                DocStringArgument = null,
+                Comments = new List<Comment>()
```
Comment (dirkrombauts) [inline src/Pickles/Pickles.Test/DocumentationBuilders/HTML/WhenFormattingStep.cs]: the `()` is unnecessary.

File: src/Pickles/Pickles.Test/DocumentationBuilders/HTML/WhenFormattingStep.cs [hunk 13]
```diff
@@ -181,5 +181,123 @@ public void Tables_are_formatted_as_list_items_with_tables_internal()
...
+                {
+                    new Comment()
```
Comment (dirkrombauts) [inline src/Pickles/Pickles.Test/DocumentationBuilders/HTML/WhenFormattingStep.cs]: the `()` is unnecessary.

File: src/Pickles/Pickles.Test/DocumentationBuilders/HTML/WhenFormattingStep.cs [hunk 14]
```diff
@@ -181,5 +181,123 @@ public void Tables_are_formatted_as_list_items_with_tables_internal()
...
+                    {
+                        Text = "    # A simple comment",
+                        Type = CommentType.StepComment
+                    },
+                    new Comment()
```
Comment (dirkrombauts) [inline src/Pickles/Pickles.Test/DocumentationBuilders/HTML/WhenFormattingStep.cs]: the `()` is unnecessary.

File: src/Pickles/Pickles.Test/DocumentationBuilders/HTML/WhenFormattingStep.cs [hunk 15]
```diff
@@ -181,5 +181,123 @@ public void Tables_are_formatted_as_list_items_with_tables_internal()
...
+        public void Multiline_comments_are_displayed_in_the_same_element()
+        {
+            var step = new Step
+            {
+                Keyword = Keyword.Given,
+                NativeKeyword = "Given ",
+                Name = "a simple step",
+                TableArgument = null,
+                DocStringArgument = null,
+                Comments = new List<Comment>()
```
Comment (dirkrombauts) [inline src/Pickles/Pickles.Test/DocumentationBuilders/HTML/WhenFormattingStep.cs]: the `()` is unnecessary.

Comment (dirkrombauts) [inline src/Pickles/Pickles.Test/DocumentationBuilders/HTML/WhenFormattingStep.cs]: the `()` is unnecessary.

File: src/Pickles/Pickles.Test/DocumentationBuilders/HTML/WhenFormattingStep.cs [hunk 16]
```diff
@@ -181,5 +181,123 @@ public void Tables_are_formatted_as_list_items_with_tables_internal()
...
+                {
+                    new Comment()
```
Comment (dirkrombauts) [inline src/Pickles/Pickles.Test/DocumentationBuilders/HTML/WhenFormattingStep.cs]: the `()` is unnecessary.

File: src/Pickles/Pickles.Test/DocumentationBuilders/HTML/WhenFormattingStep.cs [hunk 17]
```diff
@@ -181,5 +181,123 @@ public void Tables_are_formatted_as_list_items_with_tables_internal()
...
+                    {
+                        Text = "    # A first line",
+                        Type = CommentType.StepComment
+                    },
+                    new Comment()
```
Comment (dirkrombauts) [inline src/Pickles/Pickles.Test/DocumentationBuilders/HTML/WhenFormattingStep.cs]: the `()` is unnecessary.

File: src/Pickles/Pickles/DocumentationBuilders/HTML/HtmlStepFormatter.cs [hunk 18]
```diff
@@ -19,8 +19,11 @@
 //  --------------------------------------------------------------------------------------------------------------------
 
 using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
 using System.Xml.Linq;
-
+using DocumentFormat.OpenXml.Office.Word;
```
Comment (dirkrombauts) [inline src/Pickles/Pickles/DocumentationBuilders/HTML/HtmlStepFormatter.cs]: Please remove this unused using - especially since Word and Html are different things :-)

File: src/Pickles/Pickles/DocumentationBuilders/HTML/HtmlStepFormatter.cs [hunk 19]
```diff
@@ -19,8 +19,11 @@
 //  --------------------------------------------------------------------------------------------------------------------
 
 using System;
+using System.Collections.Generic;
+using System.Linq;
```
Comment (dirkrombauts) [inline src/Pickles/Pickles/DocumentationBuilders/HTML/HtmlStepFormatter.cs]: Please remove this unused using

File: src/Pickles/Pickles/DocumentationBuilders/HTML/HtmlStepFormatter.cs [hunk 20]
```diff
@@ -19,8 +19,11 @@
 //  --------------------------------------------------------------------------------------------------------------------
 
 using System;
+using System.Collections.Generic;
```
Comment (dirkrombauts) [inline src/Pickles/Pickles/DocumentationBuilders/HTML/HtmlStepFormatter.cs]: Please remove this unused using

File: src/Pickles/Pickles/ObjectModel/Mapper.cs [hunk 21]
```diff
@@ -38,7 +38,7 @@ public Mapper(string featureLanguage = LanguageServices.DefaultLanguage)
             this.mapper = new MappingEngine(configurationStore);
 
             configurationStore.CreateMap<string, Keyword>().ConvertUsing(new KeywordResolver(featureLanguage));
-
+
```
Comment (dirkrombauts) [inline src/Pickles/Pickles/ObjectModel/Mapper.cs]: Please remove the whitespace in the line

File: src/Pickles/Pickles/ObjectModel/Mapper.cs [hunk 22]
```diff
@@ -97,9 +114,42 @@ public Mapper(string featureLanguage = LanguageServices.DefaultLanguage)
...
+                            foreach (var comment in targetFeature.Comments.ToArray())
+                            {
+                                // Find the related feature
+                                var relatedFeatureElement = targetFeature.FeatureElements.LastOrDefault(x => x.Location.Line < comment.Location.Line);
+
+                                // Find the step to which the comment is related to
+                                if (relatedFeatureElement != null)
+                                {
+                                    var stepAfterComment = relatedFeatureElement.Steps.FirstOrDefault(x => x.Location.Line > comment.Location.Line);
+
```
Comment (dirkrombauts) [inline src/Pickles/Pickles/ObjectModel/Mapper.cs]: Please remove the whitespace in the line

**Comment by dirkrombauts** [comment]
Thanks for your continuing efforts!

I added a bunch of codestyle-related remarks to the files.

I would also like you to write unit tests for the mapper, like the once you find in https://github.com/picklesdoc/pickles/tree/develop/src/Pickles/Pickles.Test/ObjectModel

**Comment by ludwigjossieaux** [comment]
Hi, I've just finished the clean up + the test missing for the mapper, hopefully I didn't miss anything.

**Comment by dirkrombauts** [comment]
So far so good - thanks!

I also meant really granular tests where you test the mapping of Comment and Location in isolation, like for example the steps class mapping is being tested in https://github.com/picklesdoc/pickles/blob/develop/src/Pickles/Pickles.Test/ObjectModel/MapperTestsForStep.cs.

**Comment by ludwigjossieaux** [comment]
Ah alright, I've misunderstood the request, I've added two tests classes for comment and location, that should cover these cases

**Comment by dirkrombauts** [comment]
Thanks for your time and energy!

**Comment by dirkrombauts** [comment]
Released in version [2.6.0](https://github.com/picklesdoc/pickles/releases/tag/v2.6.0)
